include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+    for (typename parallel::distributed::Triangulation<dim>::cell_iterator it = tr->begin_active();
+         it != tr->end(); it++)
+      if (it->is_locally_owned())
+        {
+          std::vector<std::shared_ptr<DataType> > qpd = data_storage->get_data(it);
+          number_of_values = qpd[0]->number_of_values();
+          break;
+        }
+    // some processors may have no data stored, thus get the maximum among all processors:
+    number_of_values = Utilities::MPI::max(number_of_values, tr->get_communicator ());
+    Assert (number_of_values > 0,
+            ExcInternalError());
+    matrix_dofs = FullMatrix<double>(dofs_per_cell,number_of_values);
+    matrix_dofs_child = FullMatrix<double>(dofs_per_cell,number_of_values);
+    matrix_quadrature = FullMatrix<double>(n_q_points,number_of_values);
+    size = sizeof(double) * dofs_per_cell * number_of_values;
+
+    offset = tr->register_data_attach(size,
+                                      std_cxx11::bind(&ContinuousQuadratureDataTransfer<dim,DataType>::pack_function,
+                                                      this,
+                                                      std_cxx11::_1,
+                                                      std_cxx11::_2,
+                                                      std_cxx11::_3));
+  }
+
+  template<int dim, typename DataType>
+  void ContinuousQuadratureDataTransfer<dim,DataType>::interpolate ()
+  {
+    tr->notify_ready_to_unpack(offset,
+                               std_cxx11::bind(&ContinuousQuadratureDataTransfer<dim,DataType>::unpack_function,
+                                               this,
+                                               std_cxx11::_1,
+                                               std_cxx11::_2,
+                                               std_cxx11::_3));
+
+    // invalidate the pointers
+    data_storage = nullptr;
+    tr = nullptr;
+  }
+
+  template<int dim, typename DataType>
+  void ContinuousQuadratureDataTransfer<dim,DataType>::pack_function
+  (const typename parallel::distributed::Triangulation<dim,dim>::cell_iterator &cell,
+   const typename parallel::distributed::Triangulation<dim,dim>::CellStatus status,
+   void *data)
+  {
+    double *data_store = reinterpret_cast<double *>(data);
+
+    pack_cell_data(cell,data_storage,matrix_quadrature);
+
+    // project to FE
+    project_to_fe_matrix.mmult(matrix_dofs, matrix_quadrature);
+
+    std::memcpy(data_store, &matrix_dofs(0,0), size);
+  }
+
+  template<int dim, typename DataType>
+  void ContinuousQuadratureDataTransfer<dim,DataType>::
+  unpack_function (const typename parallel::distributed::Triangulation<dim,dim>::cell_iterator &cell,
+                   const typename parallel::distributed::Triangulation<dim,dim>::CellStatus status,
+                   const void *data)
+  {
+    if (status==parallel::distributed::Triangulation<dim,dim>::CELL_COARSEN)
+      {
+        Assert(false, ExcNotImplemented());
```

Comment:
had to make it a two-liner as `Assert` does not like any template arguments inside:
```
error: too many arguments provided to function-like macro invocation
```